File: EVSDKDemo/Streamer/EVDemoLiveViewController.m
//
//  EVDemoLiveViewController.m
//  EVSDKDemo
//
//  Created by mashuaiwei on 16/7/28.
//  Copyright © 2016年 cloudfocous. All rights reserved.
//

#import "EVDemoLiveViewController.h"
//#import "EVStreamer.h"
#import <EVMediaFramework/EVStreamer.h>
#import "CCAlertManager.h"
#import <AVFoundation/AVFoundation.h>
#import "EVNetRequestManager.h"
//#import "EVSDKManager.h"
#import <EVSDKBaseFramework/EVSDKManager.h>
#import "EVGenerateQRViewController.h"
#import "EVConfigBeautyLevelViewController.h"

static NSString * const kAgoraId = @"";

@interface EVDemoLiveViewController ()<EVStreamerDelegate>

@property (weak, nonatomic) IBOutlet UILabel *vidLbl;
@property (weak, nonatomic) IBOutlet UIButton *muteBtn;
@property (weak, nonatomic) IBOutlet UIButton *frontCameraBtn;
@property (weak, nonatomic) IBOutlet UIButton *flashlightBtn;
@property (weak, nonatomic) IBOutlet UIButton *beautyBtn;
@property (weak, nonatomic) IBOutlet UIButton *musicBtn;
@property (weak, nonatomic) IBOutlet UIButton *chatBtn;
@property (nonatomic, strong) EVStreamer *streamer; /**< 直播端推流器 */
@property (nonatomic, copy) NSString *vid;  /**< 视频id */
@property (nonatomic, assign) NSInteger currentBeautyLevel;
@property (nonatomic, assign) BOOL isBeautyEnable;

@property (nonatomic, assign) BOOL musicNeedContinue;   /**< 记录音乐是否需要继续播放 */

@end

@implementation EVDemoLiveViewController

#pragma mark - life circle

- (void)viewDidLoad {
    [super viewDidLoad];
    [[UIApplication sharedApplication] setIdleTimerDisabled:YES];
    
    [self configCustomUI];
    [self setupEncoder];
    [self setUpNotification];
}

- (void)dealloc{
    _streamer.delegate = nil;
    [_streamer shutDown];
    _streamer = nil;
    [[UIApplication sharedApplication] setIdleTimerDisabled:NO];
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}


#pragma mark - IBAction

- (IBAction)dismiss:(UIButton *)sender
{
    __weak typeof(self) wSelf = self;
    NSLog(@"chat state: %ld", self.streamer.chatState);
    if (self.streamer.chatState == EVVideoChatStateRemoteVideoRendered) {
        [[CCAlertManager shareInstance] performComfirmTitle:@"提示" message:@"请先结束连麦" comfirmTitle:@"好的" WithComfirm:^{
            __strong typeof(wSelf) sSelf = wSelf;
            [sSelf.streamer leaveChannel];
            sSelf.chatBtn.selected = !sSelf.chatBtn.selected;
        }];
        
        return;
    }
    
    
    [[CCAlertManager shareInstance] performComfirmTitle:@"提示" message:@"是否停止当前直播?" cancelButtonTitle:@"取消" comfirmTitle:@"确定" WithComfirm:^{
        __strong typeof(wSelf) sSelf = wSelf;
        [sSelf.streamer shutDown];
        [wSelf dismissViewControllerAnimated:YES completion:nil];
    } cancel:^{
        
    }];
}

- (IBAction)switchCamera:(UIButton *)sender
{
    BOOL font = !sender.selected;
    if (font) {
        self.flashlightBtn.selected = NO;
    }
    sender.userInteractionEnabled = NO;
    sender.selected = [self.streamer swithCamera] ? font : !font;
    sender.userInteractionEnabled = YES;
}

- (IBAction)flashLightON:(UIButton *)sender
{
    if (self.frontCameraBtn.selected) {
        return;
    }
    BOOL on = !sender.selected;
    [self.streamer turnOnFlashLight:on];
    sender.selected = on;
}

- (IBAction)mute:(UIButton *)sender {
    BOOL on = !sender.selected;
    [self.streamer muteStream:on];
    sender.selected = on;
}
- (IBAction)beauty:(UIButton *)sender {
    EVConfigBeautyLevelViewController *beautyVC = [EVConfigBeautyLevelViewController instanceVC];
    beautyVC.index = self.currentBeautyLevel-1;
    beautyVC.segValueChanged = ^(NSInteger index) {
        self.currentBeautyLevel = index;
        // 设置推流器美颜等级
        [self.streamer configBeautyLevel:index];
    };
    beautyVC.isBeautyEnable = self.isBeautyEnable;
    beautyVC.switchValueChanged = ^(BOOL YorN) {
        // 是否开启推流器美颜
        [self.streamer enableFaceBeauty:YorN];
        self.isBeautyEnable = YorN;
    };
    [beautyVC showInViewController:self];
}
- (IBAction)music:(UIButton *)sender {
    BOOL on = !sender.selected;
    if (on) {
        self.musicNeedContinue = YES;
        [self.streamer BGMPlayWithPath:[[NSBundle mainBundle] pathForResource:@"zebra" ofType:@"mp3"]];
    } else {
        self.musicNeedContinue = NO;
        [self.streamer BGMPause];
    }
    
    sender.selected = on;
}
- (IBAction)changeVolume:(UISlider *)sender {
    self.streamer.BGMVolume = sender.value;
}

- (IBAction)valueChanged:(UISlider *)sender
{
    [self.streamer cameraZoomWithFactor:sender.value];
}
- (IBAction)interactiveLiveButtonClicked:(UIButton *)sender {
    if (!self.streamer.agoraAppid) {
        [[CCAlertManager shareInstance] performComfirmTitle:nil message:@"请先设置agoraAppid" comfirmTitle:@"确定" WithComfirm:nil];
        return;
    }
    sender.selected = !sender.selected;
    if (sender.selected) {
        [self.streamer joinChannel:@"evtest002"];
    } else {
        [self.streamer leaveChannel];
    }
}
- (void)generateQR:(UIGestureRecognizer *)sender {
    if (self.vid) {
        EVGenerateQRViewController *QRVC = [EVGenerateQRViewController new];
        QRVC.infoString = self.vid;
        [self presentViewController:QRVC animated:YES completion:nil];
    }
}


#pragma mark - EVStreamerDelegate

- (void)EVStreamerStreamStateChanged:(EVVideoStreamerState)state{
    switch (state) {
        case EVVideoStreamerStreamIdle:
            NSLog(@"闲置状态，未推流");
            break;
            
        case EVVideoStreamerStreamError:
            NSLog(@"推流失败");
            break;
            
        case EVVideoStreamerStreamConnecting:
            NSLog(@"服务器连接中");
            break;
            
        case EVVideoStreamerStreamConnected:
            NSLog(@"开始推流");
            [[CCAlertManager shareInstance] performComfirmTitle:nil message:@"开始推流" comfirmTitle:@"确定" WithComfirm:nil];
            break;
            
        case EVVideoStreamerStreamDisconnecting:
            NSLog(@"断开连接");
            break;
            
        default:
            break;
    }
}

- (void)EVStreamerBufferStateChanged:(EVVideoStreamerStreamBufferState)state{
    switch (state) {
        case EVVideoStreamerStreamBufferStateNormal:
            NSLog(@"正常推流");
            break;
            
        case EVVideoStreamerStreamBufferStateLv1:
            NSLog(@"当前网络比较差，等级1，还可以坚持继续播");
            break;
            
        case EVVideoStreamerStreamBufferStateLv2:
            NSLog(@"当前网络弱爆了，等级2，建议关闭直播");
            break;
            
        case EVVideoStreamerStreamBufferStateLv3:
            NSLog(@"当前网络无法直播，可以直接掐掉了");
            break;
            
        default:
            break;
    }
}

- (void)EVStreamerUpdateVideoChatState:(EVVideoChatState)chatState{
    NSLog(@"连麦状态：%lu", chatState);
}



#pragma mark - notification

- (void)didEnterForeground
{
    if (self.musicNeedContinue) {
        [self.streamer BGMResume];
    }
    
}

- (void)didEnterBackground
{
    if (self.musicNeedContinue) {
        [self.streamer BGMPause];
    }
}


#pragma mark - private methods

- (void)configCustomUI{
    self.muteBtn.selected = [self.streamerParams[kMuteKey] boolValue];
    self.frontCameraBtn.selected = [self.streamerParams[kFrontCameraKey] boolValue];
    self.flashlightBtn.selected = [self.streamerParams[kFlashlightKey] boolValue];
    self.isBeautyEnable = [self.streamerParams[kBeautyKey] boolValue];
    if (self.isBeautyEnable) {
        self.currentBeautyLevel = 3;
    }
    [self.vidLbl addGestureRecognizer:({
        UITapGestureRecognizer *tap = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(generateQR:)];
        tap;
    })];
}

- (void)setupEncoder
{
    [self.view insertSubview:self.streamer.preview atIndex:0];
    self.streamer.delegate = self;
    self.streamer.videoBitrate = [(NSNumber *)self.streamerParams[kInitialBitrateKey] integerValue];
    self.streamer.audioBitrate = [(NSNumber *)self.streamerParams[kAudioBitrateKey] integerValue];
    self.streamer.useHEAAC = [(NSNumber *)self.streamerParams[kUseHighAACKey] boolValue];
    self.streamer.mute = [self.streamerParams[kMuteKey] boolValue];
    self.streamer.frontCamera = [self.streamerParams[kFrontCameraKey] boolValue];
    self.streamer.flashOn = [self.streamerParams[kFlashlightKey] boolValue];
    [self.streamer enableFaceBeauty:[self.streamerParams[kBeautyKey] boolValue]];
    self.streamer.agoraAppid = kAgoraId;
    
    __weak typeof(self) wSelf = self;
    [self.streamer livePrepareComplete:^(EVStreamerResponseCode responseCode, NSDictionary *result, NSError *err) {
        __strong typeof(wSelf) sSelf = wSelf;
        if (responseCode == EVStreamerResponse_Okay) {
            [sSelf startLive];
        } else {
            [[CCAlertManager shareInstance] performComfirmTitle:@"err" message:err.description comfirmTitle:@"okay" WithComfirm:nil];
        }
    }];
}

- (void)startLive{
    [self.streamer startPreview];
    
    [self generateLid];
}

- (void)generateLid{
//    NSString *url = [NSString stringWithFormat:@"http://video.api.easyvaas.com/genstream?appid=%@", [EVSDKManager appID]];
    NSString *url = [NSString stringWithFormat:@"http://api.video.easyvaas.com/v2/server/live/create?appid=%@", [EVSDKManager appID]];
    __weak typeof(self) wSelf = self;
    [EVNetRequestManager httpGetWithUrl:url moreHeaders:nil success:^(NSDictionary *info) {
        NSLog(@"------get vid info: %@", info);
        
        __strong typeof(wSelf) sSelf = wSelf;
        NSNumber *state = info[@"state"];
        if ([state integerValue] == 0) {
            NSDictionary *content = info[@"content"];
            NSString *lid = content[@"lid"];
            if (lid) {
                [sSelf startWithVid:lid];
            }
        }else{
            [[CCAlertManager shareInstance] performComfirmTitle:nil message:@"获取lid失败" cancelButtonTitle:@"确定" comfirmTitle:@"重试" WithComfirm:^{
                [sSelf generateLid];
            } cancel:nil];
        }
    } failure:^(NSError *err) {
        __strong typeof(wSelf) sSelf = wSelf;
        [[CCAlertManager shareInstance] performComfirmTitle:nil message:@"获取vid失败" cancelButtonTitle:@"确定" comfirmTitle:@"重试" WithComfirm:^{
            [sSelf generateLid];
        } cancel:nil];
    }];
}

- (void)startWithVid:(NSString *)vid {
    self.vid = vid;
    self.vidLbl.text = [NSString stringWithFormat:@"vid: %@", self.vid];
//    NSString *encodeKey = [self.key stringByAddingPercentEncodingWithAllowedCharacters:[NSCharacterSet URLQueryAllowedCharacterSet]];
//    self.streamer.URI = [NSString stringWithFormat:@"pushlocate?support=1&proto=2&lid=%@&key=%@", self.vid, encodeKey];
    self.streamer.lid = self.vid;
    __weak typeof(self) wSelf = self;
    [self.streamer liveStartComplete:^(EVStreamerResponseCode responseCode, NSDictionary *result, NSError *err) {
        __strong typeof(wSelf) sSelf = wSelf;
        if (responseCode == EVStreamerResponse_Okay) {
            [sSelf.streamer startStream];
        } else {
          [[CCAlertManager shareInstance] performComfirmTitle:nil message:[NSString stringWithFormat:@"%@", err.userInfo[@"info"]] comfirmTitle:@"OK" WithComfirm:nil];
        }
    }];
}

- (void)setUpNotification
{
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(didEnterForeground)
                                                 name:UIApplicationWillEnterForegroundNotification object:nil];
    
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(didEnterBackground)
                                                 name:UIApplicationDidEnterBackgroundNotification object:nil];
}


#pragma mark - setters and getters

- (EVStreamer *)streamer{
    if ( !_streamer ){
        _streamer = [[EVStreamer alloc] init];
    }
    
    return _streamer;
}


@end

File: EVSDKDemo_Swift/Streamer/EVDemoLiveViewController.swift
//
//  EVDemoLiveViewController.swift
//  EVSDKDemo
//
//  Created by Lcrnice on 2017/8/10.
//  Copyright © 2017年 cloudfocous. All rights reserved.
//

import UIKit
import EVMediaFramework
import EVSDKBaseFramework

class EVDemoLiveViewController: UIViewController {
  
  /// public
  var videoBitrate: UInt = 0
  var audioBitrate: EVStreamerAudioBitrate = ._48
  var mute: Bool = false
  var frontCamera: Bool = false
  var beauty: Bool = false

  /// IBOutlets
  @IBOutlet weak var vidLbl: UILabel!
  @IBOutlet weak var chatBtn: UIButton!
  @IBOutlet weak var muteBtn: UIButton!
  @IBOutlet weak var frontCameraBtn: UIButton!
  @IBOutlet weak var flashlightBtn: UIButton!
  @IBOutlet weak var beautyBtn: UIButton!
  @IBOutlet weak var musicBtn: UIButton!
  
  /// private vars
  fileprivate var streamer: EVStreamer!
  fileprivate var vid: String?
  fileprivate var currentBeautyLevel: Int = 3
  fileprivate var isBeautyEnable: Bool = true
  fileprivate var musicNeedContinue: Bool = false
  fileprivate static let kAgoraId = ""
  
  deinit {
    streamer.shutDown()
    streamer.delegate = nil
    UIApplication.shared.isIdleTimerDisabled = true
    NotificationCenter.default.removeObserver(self)
  }
  
  override func viewDidLoad() {
    super.viewDidLoad()
    
    self.configUI()
    self.setupEncoder()
    self.addLiveObserver()
  }
  
  override func didReceiveMemoryWarning() {
    super.didReceiveMemoryWarning()
  }
  
  @IBAction func interactiveLive(_ sender: UIButton) {
    guard streamer.agoraAppid != nil && streamer.agoraAppid.isEmpty == false else {
      CCAlertManager.shareInstance().performComfirmTitle("提示", message: "请先设置agoraAppid", comfirmTitle: "OK", withComfirm: nil)
      return
    }
    
    sender.isSelected = !sender.isSelected
    
    if sender.isSelected {
      streamer.joinChannel("evtest002")
    } else {
      streamer.leaveChannel()
    }
  }
  
  @IBAction func mute(_ sender: UIButton) {
    sender.isSelected = !sender.isSelected
    streamer.muteStream(sender.isSelected)
  }
  
  @IBAction func switchCamera(_ sender: UIButton) {
    let front = !sender.isSelected
    if front {
      flashlightBtn.isSelected = false
    }
    
    sender.isUserInteractionEnabled = false
    sender.isSelected = streamer.swithCamera() ? front : !front
    sender.isUserInteractionEnabled = true
  }
  
  @IBAction func switchLight(_ sender: UIButton) {
    if frontCameraBtn.isSelected {
      return
    }
    
    sender.isSelected = !sender.isSelected
    streamer.turn(onFlashLight: sender.isSelected)
  }
  
  @IBAction func switchBeauty(_ sender: UIButton) {
    let configBeautyVC = EVConfigBeautyLevelViewController.instanceVC()
    configBeautyVC.index = currentBeautyLevel - 1
    configBeautyVC.isBeautyEnable = isBeautyEnable
    configBeautyVC.segValueChanged = { [weak self] (index: Int) in
      guard let sSelf = self else {
        return
      }
      sSelf.currentBeautyLevel = index
      sSelf.streamer.configBeautyLevel(index)
    }
    configBeautyVC.switchValueChanged = { [weak self] (on: Bool) in
      guard let sSelf = self else {
        return
      }
      sSelf.streamer.enableFaceBeauty(on)
      sSelf.isBeautyEnable = on
    }
    configBeautyVC.showInViewController(parentViewController: self)
  }
  
  @IBAction func switchMusic(_ sender: UIButton) {
    sender.isSelected = !sender.isSelected
    musicNeedContinue = sender.isSelected
    
    if sender.isSelected {
      let path = Bundle.main.path(forResource: "zebra", ofType: "mp3")
      streamer.bgmPlay(withPath: path)
    } else {
      streamer.bgmPause()
    }
  }
  
  @IBAction func changeVolume(_ sender: UISlider) {
    streamer.bgmVolume = sender.value
  }
  
  @IBAction func generateQR(_ sender: Any) {
    guard vidLbl.text != nil else {
      print("没有lid")
      return
    }
    
    let qrVC = EVGenerateQRViewController()
    qrVC.infoString = vidLbl.text
    self.present(qrVC, animated: true, completion: nil)
  }
  
  @IBAction func close(_ sender: Any) {
    if streamer.chatState == .remoteVideoRendered {
      CCAlertManager.shareInstance().performComfirmTitle("提示", message: "请先结束连麦", cancelButtonTitle: "取消", comfirmTitle: "确定", withComfirm: { [weak self] in
        guard let sSelf = self else {
          return
        }
        
        sSelf.streamer.leaveChannel()
        sSelf.chatBtn.isSelected = !sSelf.chatBtn.isSelected
      }, cancel: nil)
      
      return
    }
    
    CCAlertManager.shareInstance().performComfirmTitle("提示", message: "是否停止当前直播?", cancelButtonTitle: "取消", comfirmTitle: "确定", withComfirm: { [weak self] in
      guard let sSelf = self else {
        return
      }
      
      sSelf.streamer.shutDown()
      sSelf.dismiss(animated: true, completion: nil)
    }, cancel: nil)
  }
  
}

extension EVDemoLiveViewController {
  fileprivate func setupEncoder() {
    UIApplication.shared.isIdleTimerDisabled = false
    
    streamer = EVStreamer()
    view.insertSubview(streamer.preview, at: 0)
    streamer.delegate = self as EVStreamerDelegate
    streamer.videoBitrate = videoBitrate
    streamer.audioBitrate = audioBitrate
    streamer.mute = mute
    streamer.frontCamera = frontCamera
    streamer.enableFaceBeauty(beauty)
    streamer.agoraAppid = EVDemoLiveViewController.kAgoraId
    
    streamer.livePrepareComplete { [weak self] (code: EVStreamerResponseCode, result: [AnyHashable : Any]?, error: Error?) in
      guard let sSelf = self else {
        return
      }
      
      if code == .okay {
        sSelf.streamer.startPreview()
        sSelf.generateLid()
      } else {
        CCAlertManager.shareInstance().performComfirmTitle("开播失败", message: error?.localizedDescription, comfirmTitle: "OK", withComfirm: nil)
      }
    }
  }
  
  fileprivate func configUI() {
    muteBtn.isSelected = mute
    frontCameraBtn.isSelected = frontCamera
    isBeautyEnable = beauty
  }
  
  fileprivate func addLiveObserver() {
    NotificationCenter.default.addObserver(forName: NSNotification.Name.UIApplicationWillEnterForeground, object: nil, queue: nil, using: { [weak self] _ in
      guard let sSelf = self else {
        return
      }
      
      if sSelf.musicNeedContinue {
        sSelf.streamer.bgmResume()
      }
    })
    
    NotificationCenter.default.addObserver(forName: NSNotification.Name.UIApplicationDidEnterBackground, object: nil, queue: nil, using: { [weak self] _ in
      guard let sSelf = self else {
        return
      }
      
      if sSelf.musicNeedContinue {
        sSelf.streamer.bgmPause()
      }
    })
  }
  
  private func generateLid() {
    let appID = EVSDKManager.appID()
    let url = "http://api.video.easyvaas.com/v2/server/live/create?appid=\(appID)"
    EVNetRequestManager.httpGet(withUrl: url, moreHeaders: nil, success: { [weak self] (info: [AnyHashable : Any]?) in
      guard let sSelf = self else {
        return
      }
      
      print("generate lid info:\(String(describing: info))")
      let code: Int = info?["state"] as! Int
      if code == 0 {
        let content = info?["content"] as! [String : String]
        let lid = content["lid"]
        
        if lid?.isEmpty == false && lid != nil {
          self?.startLive(lid: lid!)
        }
      } else {
        sSelf.reloadLid()
      }
    }) { [weak self] (error: Error?) in
      guard let sSelf = self else {
        return
      }
      sSelf.reloadLid()
    }
  }
  
  private func reloadLid() {
    CCAlertManager.shareInstance().performComfirmTitle(nil, message: "获取lid失败", cancelButtonTitle: "确定", comfirmTitle: "重试", withComfirm: { 
      self.generateLid()
    }, cancel: nil)
  }
  
  private func startLive(lid: String) {
    vidLbl.text = lid
    streamer.lid = lid
    
    streamer.liveStartComplete { [weak self] (code: EVStreamerResponseCode, result: [AnyHashable : Any]?, error: Error?) in
      guard let sSelf = self else {
        return
      }
      
      if code == .okay {
        sSelf.streamer.startStream()
      } else {
        print("start live failed.")
      }
    }
  }
  
}


// MARK: - EVStreamerDelegate
extension EVDemoLiveViewController: EVStreamerDelegate {
  func evStreamerStreamStateChanged(_ state: EVVideoStreamerState) {
    switch state {
    case .streamIdle:
      print("闲置状态，未推流")
    case .streamConnecting:
      print("服务器连接中")
    case .streamConnected:
      print("开始推流")
      CCAlertManager.shareInstance().performComfirmTitle("开始推流", message: nil, comfirmTitle: "OK", withComfirm: nil)
    case .streamDisconnecting:
      print("断开连接")
    case .streamError:
      print("推流失败")
    }
  }
  
  func evStreamerBufferStateChanged(_ state: EVVideoStreamerStreamBufferState) {
    switch state {
    case .normal:
      print("正常推流")
    case .lv1:
      print("当前网络比较差，等级1，还可以坚持继续播")
    case .lv2:
      print("当前网络弱爆了，等级2，建议关闭直播")
    case .lv3:
      print("当前网络无法直播，可以直接掐掉了")
    }
  }
  
  func evStreamerUpdate(_ chatState: EVVideoChatState) {
    print("连麦状态:\(chatState)")
  }
  
}

